File: apps/web/src/service/api/analysis-compute.ts
import type { AnalysisCounts, AnalysisPipeline, AnalysisPreview, AnalysisRun, AnalysisSelection, AnalysisSource } from "./analysis"
import type { ComputeResourceLimits, ResearchComputeEnvironment } from "./research-compute"
import type { ComputeOutputDraft, ResearchComputeJob } from "./research-compute-jobs"
import { request } from "../request"

export interface AnalysisComputeInputFile {
  input_id: string
  field_path: ["var", string]
}

export interface AnalysisComputeInputFileField {
  field_path: ["var", string]
  title: string
  file_extensions: string[] | null
  nullable: boolean
}

export interface AnalysisComputeInputFileLimits {
  max_files: number
  max_file_bytes: number
  max_total_bytes: number
  manifest_filename: "attachments.json"
}

export interface AnalysisComputeInputFileReceipt extends AnalysisComputeInputFile {
  record_id: string
  record_version: number
  record_hash: string
  protocol_version: string
  file_id: string
  filename: string
  media_type: string
  byte_size: number
  checksum_sha256: string
  mount_name: string
  file_metadata_digest: string
}

export interface AnalysisComputeInputFilesEnvelope {
  schema: string
  count: number
  total_bytes: number
  files: AnalysisComputeInputFileReceipt[]
  manifest: { filename: string, byte_size: number, checksum_sha256: string }
}

export interface AnalysisComputeRecipe {
  kind: "compute"
  environment_revision_id: string
  language: "python" | "r"
  source_code: string
  parameters: Record<string, unknown>
  output_files: ComputeOutputDraft[]
  input_files?: AnalysisComputeInputFile[]
}

export interface AnalysisComputeEnvironment {
  id: string
  revision_id: string
  revision: number
  name: string
  image_ref: string
  allowed_languages: Array<"python" | "r">
  resource_limits: ComputeResourceLimits
  network_policy: "none" | "egress_allowlist"
  allowed_egress_hosts: string[]
  input_schema: Record<string, unknown>
  result_schema: Record<string, unknown>
  estimated_cost: string | null
  currency: string | null
  authorized_runner_count: number
  ready_runner_count: number
}

export interface AnalysisComputeContext {
  environments: AnalysisComputeEnvironment[]
  approvers: Array<{ id: string, name: string }>
  source: { record_count: number, source_digest: string, filename: "records.json" }
  max_source_bytes: number
  input_file_fields?: AnalysisComputeInputFileField[]
  input_file_limits?: AnalysisComputeInputFileLimits
}

export interface AnalysisComputePreviewRequest {
  protocol_id: string
  selection: AnalysisSelection
  question: string
  recipe: AnalysisComputeRecipe
  approver_user_id: string
  max_cost?: string
  budget_currency?: string
  deadline_at?: string
  pipeline_revision_id?: string
  rerun_of_id?: string
  ai_draft_id?: string
}

export interface AnalysisComputeContract {
  environment: ResearchComputeEnvironment
  source: { language: "python" | "r", code: string, sha256: string, bytes: number }
  input: { record_count: number, bytes: number, sha256: string, filename: string }
  parameters: Record<string, unknown>
  output_files: ComputeOutputDraft[]
  input_files?: AnalysisComputeInputFilesEnvelope
  approver: { id: string, name: string }
  cost: { estimated_cost: string | null, currency: string | null, max_cost: string | null, budget_currency: string | null }
  deadline_at?: string | null
  authorized_runner_count: number
  ready_runner_count: number
  approval_required: true
}

export interface AnalysisComputePreview extends Omit<AnalysisPreview, "recipe" | "summary"> {
  recipe: AnalysisComputeRecipe
  summary: {
    compute: AnalysisComputeContract
    counts: AnalysisCounts
    sources: AnalysisSource[]
    source_digest: string
    protocol_name: string
    project_name: string
    visibility: "private"
  }
}

export interface AnalysisComputeApproval {
  state: "pending" | "approved" | "rejected" | "cancelled"
  revision: number
  approver_user_id: string
  can_approve: boolean
  contract_digest: string
  reason?: string
  decided_at?: string | null
}

export interface AnalysisComputeDetail {
  contract: AnalysisComputeContract
  job: Omit<ResearchComputeJob, "output_manifest"> & { output_manifest: AnalysisComputeOutput[] }
  approval: AnalysisComputeApproval
  events: Array<{ id: string, kind: string, created_at: string, details?: Record<string, unknown> }>
  run: AnalysisRun
}

export interface AnalysisComputeOutput {
  id: string
  mount_name: string
  asset_name: string
  media_type: string
  max_bytes: number
  byte_size: number | null
  checksum_sha256: string | null
  required: boolean
  status: string
}

export interface AnalysisComputeResult {
  computed_result: Record<string, unknown>
  outputs: AnalysisComputeOutput[]
  usage: Record<string, number>
  actual_cost: string | null
  currency: string | null
  input_files?: AnalysisComputeInputFilesEnvelope
}

export interface AnalysisComputeSeed {
  recipe: AnalysisComputeRecipe
  pipeline?: AnalysisPipeline
  pipelineRevisionId?: string
  rerunOfId?: string
}

async function requiredData<T>(options: Parameters<typeof request<T>>[0]) {
  const { data, error } = await request<T>({ ...options, metadata: { showError: false } })
  if (error)
    throw error
  if (data === null)
    throw new Error("Analysis Compute service returned no data")
  return data
}

export function fetchAnalysisComputeContext(protocolId: string, selection: AnalysisSelection) {
  return requiredData<AnalysisComputeContext>({ url: `/protocols/${protocolId}/analysis-compute-context`, method: "POST", data: selection })
}

export function previewAnalysisCompute(payload: AnalysisComputePreviewRequest) {
  return requiredData<AnalysisComputePreview>({ url: "/analyses/compute/preview", method: "POST", data: payload })
}

export function confirmAnalysisCompute(payload: { preview_id: string, preview_digest: string, client_idempotency_key: string }) {
  return requiredData<AnalysisRun>({ url: "/analyses/compute", method: "POST", data: payload })
}

export function fetchAnalysisCompute(id: string, approvalOnly = false) {
  return requiredData<AnalysisComputeDetail>({ url: `/analyses/${id}/${approvalOnly ? "compute-approval" : "compute"}` })
}

export function decideAnalysisCompute(id: string, payload: { decision: "approved" | "rejected", expected_revision: number, contract_digest: string, reason: string }) {
  return requiredData<AnalysisComputeDetail>({ url: `/analyses/${id}/compute-approval`, method: "POST", data: payload })
}

export function cancelAnalysisCompute(id: string, payload: { expected_revision: number, reason: string, contract_digest: string }) {
  return requiredData<{ analysis_id: string, job_id: string, status: ResearchComputeJob["status"], revision: number }>({ url: `/analyses/${id}/compute-cancel`, method: "POST", data: payload })
}

export async function downloadAnalysisComputeOutput(id: string, outputId: string) {
  const { data, error } = await request<Blob, "blob">({ url: `/analyses/${id}/compute/outputs/${outputId}`, responseType: "blob", metadata: { showError: false } })
  if (error)
    throw error
  if (data === null)
    throw new Error("Analysis output download returned no data")
  return data
}

export function fetchAnalysisComputeApprovals(projectId: string) {
  return requiredData<{ items: Array<{ analysis_id: string, question: string, requested_at: string, approval_state: string, approver_user_id: string }> }>({ url: "/analysis-compute-approvals", params: { project_id: projectId } })
}

File: apps/api/app/routers/analysis_compute.py
"""Private analysis adapter for the shared, approval-gated Compute runtime."""

from urllib.parse import quote
from uuid import UUID, uuid4

from fastapi import APIRouter, HTTPException, Query, Response
from fastapi.responses import StreamingResponse
from sqlalchemy import select

from app.database import DBSession
from app.libs.file_storage import get_file_with_stream
from app.models.analysis import AnalysisRun
from app.models.analysis_compute import AnalysisCompute, AnalysisComputeEvent
from app.models.knowledge import ResearchFileBlob
from app.models.research_execution import ResearchComputeJob, ResearchComputeJobOutput
from app.models.user import User
from app.routers.analyses import AnalysisConfirmRequest, run_payload
from app.routers.depends import CurrentUser
from app.services.analysis_compute import (
    analysis_compute_input_bytes,
    analysis_runner_counts,
    authorize_approver,
    authorize_compute_execution,
    compute_options,
    confirm_compute,
    emit_analysis_compute_event,
    instant,
    money,
    owned_compute,
    preview_compute,
)
from app.services.analysis_compute_contracts import (
    AnalysisComputeCancel,
    AnalysisComputeDecision,
    AnalysisComputeDraft,
)
from app.services.analysis_engine import AnalysisError
from app.services.record_analyses import AnalysisSelection, analysis_scope, utcnow
from app.services.research_compute_jobs import (
    compute_job_snapshot,
    compute_source_digest,
)

router = APIRouter(tags=["analysis"])
FINAL_STATUSES = {"completed", "failed", "cancelled"}


async def compute_detail(db, user, run, details, job, *, approval_only=False):
    """Reconstruct the sealed contract; never depend on an expiring preview."""
    import hashlib

    authorized, ready = await analysis_runner_counts(
        db, job.compute_environment_revision_id
    )
    approver = await db.get(User, details.approver_user_id)
    can_approve = False
    if (
        details.approver_user_id == user.id
        and details.approval_state == "pending"
        and job.status == "awaiting_approval"
    ):
        try:
            _, project, _ = await analysis_scope(db, run.protocol_id, user)
            await authorize_approver(
                db, run.protocol_id, project, run.source_snapshot, user
            )
            if run.recipe.get("input_files"):
                from app.services.analysis_compute_files import authorize_input_files

                await authorize_input_files(db, run, user)
            can_approve = True
        except HTTPException as exc:
            if exc.status_code not in {403, 404}:
                raise
    source = analysis_compute_input_bytes(run)
    contract = {
        "environment": job.environment_snapshot,
        "source": {
            "language": job.language,
            "code": job.source_code,
            "sha256": compute_source_digest(job.source_code),
            "bytes": len(job.source_code.encode("utf-8")),
        },
        "input": {
            "filename": "records.json",
            "record_count": len(run.source_snapshot["records"]),
            "bytes": len(source),
            "sha256": hashlib.sha256(source).hexdigest(),
        },
        "parameters": job.input_payload,
        "output_files": run.recipe["output_files"],
        "approver": {
            "id": str(details.approver_user_id),
            "name": (approver.name or approver.username) if approver else "",
        },
        "cost": {
            "estimated_cost": money(job.estimated_cost),
            "currency": job.currency,
            "max_cost": money(details.max_cost),
            "budget_currency": details.budget_currency,
        },
        "deadline_at": instant(details.deadline_at),
        "authorized_runner_count": authorized,
        "ready_runner_count": ready,
        "approval_required": True,
    }
    if getattr(details, "input_file_manifest", None):
        contract["input_files"] = details.input_file_manifest
    events = (
        await db.scalars(
            select(AnalysisComputeEvent)
            .where(
                AnalysisComputeEvent.analysis_run_id == run.id,
            )
            .order_by(
                AnalysisComputeEvent.created_at.desc(), AnalysisComputeEvent.id.desc()
            )
            .limit(100)
        )
    ).all()
    job_data = compute_job_snapshot(job, include_source=True)
    if approval_only:
        # Approval grants neither a private report nor its computed artifacts.
        job_data.update(
            result={}, output_manifest=[], usage={}, error=None, cancel_reason=None
        )
    run_data = run_payload(run, summary=approval_only)
    if approval_only:
        run_data["error"] = None
        # The adopted code is reviewed in contract.source; original questions,
        # discarded AI proposals and inherited private draft history are not.
        run_data["ai_provenance"] = {}
    return {
        "run": run_data,
        "job": job_data,
        "contract": contract,
        "approval": {
            "state": details.approval_state,
            "revision": details.approval_revision,
            "approver_user_id": details.approver_user_id,
            "can_approve": can_approve,
            "contract_digest": details.contract_digest,
            "reason": details.decision_reason,
            "decided_at": details.decided_at,
            "requested_at": details.requested_at,
        },
        "events": [
            {
                "id": event.id,
                "kind": event.kind,
                "created_at": event.created_at,
                "details": {} if approval_only else event.payload,
            }
            for event in events
        ],
    }


@router.post("/protocols/{protocol_id}/analysis-compute-context")
async def get_analysis_compute_context(
    protocol_id: UUID,
    selection: AnalysisSelection,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    try:
        return await compute_options(db_session, current_user, protocol_id, selection)
    except AnalysisError as exc:
        raise HTTPException(422, str(exc)) from exc


@router.post("/analyses/compute/preview")
async def preview_analysis_compute(
    params: AnalysisComputeDraft,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    try:
        preview = await preview_compute(db_session, params, current_user)
    except AnalysisError as exc:
        raise HTTPException(422, str(exc)) from exc
    await db_session.commit()
    return preview.as_dict()


@router.post("/analyses/compute")
async def confirm_analysis_compute(
    params: AnalysisConfirmRequest,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    try:
        run = await confirm_compute(
            db_session,
            current_user,
            preview_id=params.preview_id,
            preview_digest=params.preview_digest,
            key=params.client_idempotency_key,
        )
    except AnalysisError as exc:
        raise HTTPException(422, str(exc)) from exc
    await db_session.commit()
    return run_payload(run)


@router.get("/analyses/{analysis_id}/compute")
async def get_analysis_compute(
    analysis_id: UUID,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    run, details, job = await owned_compute(db_session, analysis_id, current_user)
    return await compute_detail(db_session, current_user, run, details, job)


@router.get("/analyses/{analysis_id}/compute-approval")
async def get_analysis_compute_approval(
    analysis_id: UUID,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    run, details, job = await owned_compute(
        db_session, analysis_id, current_user, approval_access=True
    )
    return await compute_detail(
        db_session, current_user, run, details, job, approval_only=True
    )


@router.get("/analysis-compute-approvals")
async def list_analysis_compute_approvals(
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
    project_id: UUID,
    limit: int = Query(30, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    response.headers["Cache-Control"] = "private, no-store"
    rows = (
        await db_session.execute(
            select(AnalysisRun, AnalysisCompute)
            .join(
                AnalysisCompute,
                AnalysisCompute.analysis_run_id == AnalysisRun.id,
            )
            .where(
                AnalysisRun.project_id == project_id,
                AnalysisCompute.approver_user_id == current_user.id,
                AnalysisCompute.approval_state == "pending",
            )
            .order_by(AnalysisCompute.requested_at, AnalysisRun.id)
            .offset(offset)
            .limit(limit)
        )
    ).all()
    items = []
    for run, details in rows:
        try:
            await owned_compute(db_session, run.id, current_user, approval_access=True)
        except HTTPException as exc:
            if exc.status_code in {403, 404}:
                continue
            raise
        items.append(
            {
                "analysis_id": run.id,
                "question": run.question,
                "requested_at": details.requested_at,
                "approval_state": details.approval_state,
                "approver_user_id": details.approver_user_id,
            }
        )
    return {"items": items}


@router.post("/analyses/{analysis_id}/compute-approval")
async def decide_analysis_compute(
    analysis_id: UUID,
    params: AnalysisComputeDecision,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    from app.services.workflow_compute_runtime import bound_compute

    # Authenticate the ordinary private reader before revealing a workflow binding.
    await owned_compute(db_session, analysis_id, current_user, approval_access=True)
    if await bound_compute(db_session, analysis_id):
        raise HTTPException(
            409, "Approve this Compute analysis through its Research Task"
        )
    run, details, job = await owned_compute(
        db_session, analysis_id, current_user, approval_access=True, lock=True
    )
    if details.approver_user_id != current_user.id:
        raise HTTPException(403, "Only the explicitly selected approver may decide")
    _, project, _ = await analysis_scope(db_session, run.protocol_id, current_user)
    await authorize_approver(
        db_session, run.protocol_id, project, run.source_snapshot, current_user
    )
    if params.contract_digest != details.contract_digest:
        raise HTTPException(409, "Approval contract changed; review again")
    if (
        details.approval_state == params.decision
        and details.approval_revision == params.expected_revision + 1
        and details.decision_reason == params.reason
        and details.decided_by_user_id == current_user.id
    ):
        return await compute_detail(
            db_session, current_user, run, details, job, approval_only=True
        )
    if (
        details.approval_state != "pending"
        or details.approval_revision != params.expected_revision
        or job.status != "awaiting_approval"
    ):
        raise HTTPException(409, "Approval changed; review again")
    now = utcnow()
    if params.decision == "approved":
        await authorize_compute_execution(db_session, job, require_approved=False)
        authorized, _ = await analysis_runner_counts(
            db_session, job.compute_environment_revision_id
        )
        if not authorized:
            raise HTTPException(409, "No authorized Runner supports private analysis")
        job.status = "queued"
        job.approved_at = now
        job.queued_at = now
    else:
        job.status = "cancelled"
        job.completed_at = now
        run.status = "cancelled"
        run.finished_at = now
        run.error = "Compute approval was rejected"
    details.approval_state = params.decision
    details.approval_revision += 1
    details.decided_by_user_id = current_user.id
    details.decided_at = now
    details.decision_reason = params.reason
    job.revision += 1
    await emit_analysis_compute_event(
        db_session,
        run.id,
        f"compute.{params.decision}",
        actor_user_id=current_user.id,
        payload={"reason": params.reason, "contract_digest": details.contract_digest},
        key=f"decision:{details.approval_revision}",
    )
    await db_session.commit()
    return await compute_detail(
        db_session, current_user, run, details, job, approval_only=True
    )


@router.post("/analyses/{analysis_id}/compute-cancel")
async def cancel_analysis_compute(
    analysis_id: UUID,
    params: AnalysisComputeCancel,
    db_session: DBSession,
    current_user: CurrentUser,
    response: Response,
):
    response.headers["Cache-Control"] = "private, no-store"
    owner_job = await db_session.scalar(
        select(ResearchComputeJob.id).where(
            ResearchComputeJob.analysis_run_id == analysis_id,
            ResearchComputeJob.created_by_user_id == current_user.id,
        )
    )
    if owner_job is None:
        raise HTTPException(404, "Private computation not found")
    from app.services.workflow_compute_runtime import bound_compute

    if await bound_compute(db_session, analysis_id):
        raise HTTPException(
            409, "Cancel this Compute analysis through its Research Task"
        )
    await db_session.scalar(
        select(AnalysisRun)
        .where(AnalysisRun.id == analysis_id)
        .with_for_update()
        .execution_options(populate_existing=True)
    )
    # A creator may STOP their own job even after Record access is revoked.
    # This endpoint returns only a cancellation receipt, never source/results.
    job = await db_session.scalar(
        select(ResearchComputeJob)
        .where(
            ResearchComputeJob.analysis_run_id == analysis_id,
            ResearchComputeJob.created_by_user_id == current_user.id,
        )
        .with_for_update()
        .execution_options(populate_existing=True)
    )
    details = await db_session.get(AnalysisCompute, analysis_id, populate_existing=True)
    if job is None or details is None:
        raise HTTPException(404, "Private computation not found")
    if params.contract_digest != details.contract_digest:
        raise HTTPException(409, "Cancellation contract changed; review again")
    if (
        job.status in {"cancel_requested", "cancelled"}
        and job.revision == params.expected_revision + 1
        and job.cancel_reason == params.reason
    ):
        return {
            "analysis_id": analysis_id,
            "job_id": job.id,
            "status": job.status,
            "revision": job.revision,
        }
    if job.revision != params.expected_revision or job.status in FINAL_STATUSES:
        raise HTTPException(409, "Compute Job changed; review again")
    now = utcnow()
    previous = job.status
    job.cancel_reason = params.reason
    job.cancel_requested_at = now
    job.revision += 1
    run = await db_session.get(AnalysisRun, analysis_id, populate_existing=True)
    if previous in {"awaiting_approval", "queued"}:
        job.status = "cancelled"
        job.completed_at = now
        run.status = "cancelled"
        run.finished_at = now
        if details.approval_state == "pending":
            details.approval_state = "cancelled"
            details.approval_revision += 1
    else:
        job.status = "cancel_requested"
    run.error = (
        "Compute cancellation requested"
        if job.status == "cancel_requested"
        else "Compute cancelled"
    )
    await emit_analysis_compute_event(
        db_session,
        analysis_id,
        f"compute.{job.status}",
        actor_user_id=current_user.id,
        payload={"reason": params.reason, "previous_status": previous},
        key=f"cancel:{job.revision}",
    )
    await db_session.commit()
    return {
        "analysis_id": analysis_id,
        "job_id": job.id,
        "status": job.status,
        "revision": job.revision,
    }


@router.get("/analyses/{analysis_id}/compute/outputs/{output_id}")
async def download_analysis_compute_output(
    analysis_id: UUID, output_id: UUID, db_session: DBSession, current_user: CurrentUser
):
    run, _, job = await owned_compute(db_session, analysis_id, current_user)
    return await stream_analysis_compute_output(
        db_session, current_user, run, job, output_id
    )


async def stream_analysis_compute_output(db_session, current_user, run, job, output_id):
    """Shared receipt verification; the caller owns scope and source authorization."""
    if run.status != "succeeded" or job.status != "completed":
        raise HTTPException(409, "Compute output is not a completed analysis result")
    row = (
        await db_session.execute(
            select(ResearchComputeJobOutput, ResearchFileBlob)
            .join(
                ResearchFileBlob,
                ResearchFileBlob.id == ResearchComputeJobOutput.blob_id,
            )
            .where(
                ResearchComputeJobOutput.id == output_id,
                ResearchComputeJobOutput.compute_job_id == job.id,
                ResearchComputeJobOutput.registered_at.is_not(None),
            )
        )
    ).first()
    if row is None:
        raise HTTPException(404, "Private analysis output not found")
    output, blob = row
    # Compare actual storage receipts with the immutable successful report.
    sealed = next(
        (
            item
            for item in (run.result or {}).get("outputs", [])
            if item.get("id") == str(output.id)
        ),
        None,
    )
    if (
        sealed is None
        or sealed.get("checksum_sha256") != blob.checksum_sha256
        or sealed.get("byte_size") != blob.size_bytes
    ):
        raise HTTPException(409, "Private output integrity check failed")
    await emit_analysis_compute_event(
        db_session,
        run.id,
        "compute.output_downloaded",
        actor_user_id=current_user.id,
        payload={"output_id": str(output.id), "sha256": blob.checksum_sha256},
        key=f"download:{uuid4()}",
    )
    await db_session.commit()
    return StreamingResponse(
        get_file_with_stream(blob.storage_object_key, backend=blob.storage_backend),
        media_type=blob.content_type,
        headers={
            "Cache-Control": "private, no-store",
            "X-Content-Type-Options": "nosniff",
            "Content-Disposition": f"attachment; filename*=UTF-8''{quote(output.mount_name)}",
            "Content-Length": str(blob.size_bytes),
            "X-Content-SHA256": blob.checksum_sha256,
        },
    )
